[ re #4041 ] internalising the inspect idiom

gallais opened the pull request:
fixes #4088 too

gallais committed:
[ test ] for the inspect idiom

gallais committed:
Merge branch 'master' into inspect

nad commented:
The following variant of your example works fine:
```agda
open import Agda.Builtin.Nat
open import Agda.Builtin.Equality
open import Agda.Builtin.Sigma

double : Nat → Set
double m = Σ Nat λ n → m ≡ n + n

inspect : {A : Set} (x : A) → Σ A λ y → y ≡ x
inspect x = x , refl

2*_ : Nat → Σ Nat double
2*_ n with m , eq ← inspect (n + n) = m , n , eq
```
Do you have some example where the new feature cannot easily be simulated using library code? Perhaps something that involves pattern matching?

gallais committed:
[ fix ] post-merge fix

gallais committed:
[ fix ] whitespace violation

gallais committed:
Merge branch 'master' into inspect

gallais committed:
Revert "Revert "Merge branch 'master' of github.com:agda/agda""

This reverts commit bcc1d86fba1aca784300d0e0a5a4b20e7bc122a4.

gallais committed:
[ fix ] only require refl if the idiom is used

gallais commented:
> Do you have some example where the new feature cannot easily be simulated using library code? 

Right, I think the problem with your `inspect` is that `n + n` does not get
abstracted over in the type of the RHS. I will try to concoct an example.

gallais committed:
[ test ] expanded test case for inspect

gallais committed:
[ fix ] the inspect idiom for non-irrefutable withs

gallais commented:
Bikeshedding: `f xs with eq .. e` is very ugly. It would be nice to get the parser
to accept `f xs with eq : e`. Suggestions for an alternative syntax are welcome.
Do remember that whatever is chosen needs to be compatible with `p <- e` too.

Edit: Ok, we have a plan for refactoring the grammar and maybe getting
`f xs with eq : e` to work.

gallais committed:
[ inspect ] getting the grammar in a good shape!

gallais committed:
[ fix ] whitespace violation

nad commented:
> Edit: Ok, we have a plan for refactoring the grammar and maybe getting
> `f xs with eq : e` to work.

I'd prefer to reserve `:` for type signatures. I think it makes sense to allow type signatures for `with` expressions.

gallais committed:
[ fix #4088 ] test case

gallais commented:
The logic behind this name is that when we write `eq : p <- e` we do get a proof
`eq` of type `p ≡ e`.

gallais commented:
What about `f xs inspect e as n` for what is currently written `f xs with n : e`?
This requires stealing the keyword `inspect`.

gallais commented:
More general design incoming.

nad commented:
> What about `f xs inspect e as n` for what is currently written `f xs with n : e`?

Isn't the idea that it should be possible to write something like `f xs with n : e | e′`?

andreasabel commented:
> What about `f xs inspect e as n` for what is currently written `f xs with n : e`?

Or `f xs with n inspecting e`?
And `f xs with n inspecting p <- e`?

`f xs with n witnessing e`
`f xs with n witnessing p <- e`
`f xs with n of p <- e`

Guillaume:
`f xs with e in n`       (winner)
`f xs with p <- e in n`  (winner)
(But: not to be confused with `let p <- e in n`.)
`f xs with p <- e as n`

`f xs with p <- e in-the-type-of n`
`f xs with e witnessed-in n`
`f xs with e witnessed in n`
`f xs with p <- e witnessed-in n`
`f xs with p <- e witnessed in n`

`f xs with pattern e in n`
`f xs with pattern p <- e in n`

gallais committed:
Merge branch 'master' into inspect

gallais committed:
[ syntax ] using `with e in eq` instead of `with eq : e`

The patch for the syntactic change is absolutely trivial. The
hardest thing today was merging ~2300 patches into the branch.

gallais commented:
The changes break some of @andreasabel' improved error messages when
someone tries to write a `with` clause in a type declaration. I'll
look into it later.

gallais committed:
[ std-lib ] back to same ref as master

gallais committed:
[ fix ] revive good error messages

gallais committed:
Merge branch 'master' into inspect

gallais commented:
I am being victimised by github actions. :sob: 


Edit: ready except for https://github.com/agda/agda/blob/6b04ce5c332ce7f1e8901abf5f4ed7b209767d6a/src/full/Agda/TypeChecking/Abstract.hs#L64-L70 which I guess we'll talk about on Wednesday

gallais committed:
[ done! ] make it generic in the equality type

gallais committed:
[ refactor ] only use `checkType` once

Now that `checkType` actually instantiates the metas, it is not
necessary to use `sortOf` first to figure out the sort of the
equality constraint and then `checkType` to solve the remaining
metas.

gallais committed:
Merge branch 'master' into inspect

gallais committed:
[ test ] error messages have changed

andreasabel commented on test/Succeed/with-equalities.agda:
```diff
@@ -0,0 +1,63 @@
+    -- If only we had somehow recorded that `p ← (2 * n + 1)` happened!
+    -- This is what the inspect construct allows us to do.
+
+    with p ← (2 * n + 1) in eq | xs
```
Why to we have to `with` on `xs` here, as opposed to just match on the original `xs` in the subsequent clauses?

gallais replied:
The pattern-matching `with` is a bit like a `rewrite` in that it does
not introduce new clauses.

So in this artificial example there won't be a subsequent clause because
I named `2 * n + 1` as `p`.

gallais committed:
[ fix ] flip the equality for better ergonomics

As demonstrated by `filter-filter` we often want to use the equality
to unfold a computation whose result we already know. This is annoying
if the equality is `var 0 = e` because `e` is the computation we want
to unfold and `var 0` has typically been refined to a pattern.
With the flipped equality `e = var 0` we get a much more useful assumption.

gallais committed:
[ doc ] changelog and user manual

gallais committed:
[ manual ] fix missing code block types

nad commented:
I thought the idea was that we should not merge new features until after the next release.

The documentation of this feature does not explain how the translation to the core language handles `with in`.

gallais commented:
I was not sure whether this qualified as a new feature, given it's
been around for a year and a half and is also a bugfix for #4088.

We can revert the commit and re-open this PR if necessary.
